mitiq/observable/observable.py

```diff
@@ -128,6 +153,23 @@ def expectation(
 
         return Executor(execute).evaluate(circuit, observable=self)[0]
 
+    def _combine_duplicates(self) -> None:
+        d: Dict[PauliString, PauliString] = {}
+        for pauli_string in self._paulis:
+            cache_key = pauli_string.with_coeff(1)
+            if cache_key in d:
+                new_coeff = pauli_string.coeff + d[cache_key].coeff
+                d[cache_key] = d[cache_key].with_coeff(new_coeff)
+            else:
+                d[cache_key] = pauli_string
+        self._paulis = list(
+            [
+                pauli_string
+                for pauli_string in d.values()
+                if pauli_string.coeff != 0
+            ]
+        )
```

Comment:
It is probably safer to `return` a new `Observable` based on the new combined `PauliString`s instead of mutating the internal private attribute `self._paulis`. In the first case `self.partition()` is called again by the __init__ method, in the second case it's not re-called and this seems dangerous.